deps/portal/src/Channel.cpp

```diff
@@ -17,90 +17,150 @@
-
-            const uint32_t numberOfBytesToAskFor = 65536; // (1 << 16); // This is the value in DarkLighting
-            uint32_t numberOfBytesReceived = 0;
-
-            char buffer[numberOfBytesToAskFor];
-
-            int ret = usbmuxd_recv_timeout(conn, (char *)&buffer, numberOfBytesToAskFor, &numberOfBytesReceived, 10);
-
-            if (ret == 0)
-            {
-                if (numberOfBytesReceived > 0)
-                {
-                    if (running) {
-                        protocol->processData((char *)buffer, numberOfBytesReceived);
-                    }
-                }
-            }
-            else
-            {
-                portal_log("%d: There was an error receiving data\n", conn);
-                running = false;
-            }
-        }
-    }
-
-    void Channel::simpleDataPacketProtocolDelegateDidProcessPacket(std::vector<char> packet, int type, int tag)
-    {
-        std::shared_ptr<ChannelDelegate> strongDelegate = delegate.lock();
-        if (strongDelegate) {
-            strongDelegate->channelDidReceivePacket(packet, type, tag);
-        }
-    }
+	uint32_t sentBytes = 0;
+
+	usbmuxd_send(conn, &data[0], data.size(), &sentBytes);
+}
+
+void Channel::setState(State state)
+{
+	if (state == getState()) {
+		return;
+	}
+
+	//std::cout << "Channel:setState: " << state << std::endl;
+	_state = state;
+
+	if (auto delegate = this->delegate.lock()) {
+		delegate->channelDidChangeState(state);
+	}
+}
+
+/** Returns true if the thread was successfully started, false if there was an error starting the thread */
+bool Channel::StartInternalThread()
+{
+	_thread = std::thread(InternalThreadEntryFunc, this);
+	return true;
+}
+
+/** Will not return until the internal thread has exited. */
+void Channel::WaitForInternalThreadToExit()
+{
+	//if (_thread == nullptr) {
+	//	return;
+	//}
+	std::unique_lock<std::mutex> lock(worker_mutex);
+
+	if (_thread.joinable() == true) {
+		_thread.join();
+	}
+
+	//_thread = nullptr;
+	lock.unlock();
+}
+
+void Channel::StopInternalThread()
+{
+	running = false;
+}
+
+// TODO: https://stackoverflow.com/questions/58477291/function-exceeds-stack-size-consider-moving-some-data-to-heap-c6262
+void Channel::InternalThreadEntry()
+{
+	while (running) {
+
+		const uint32_t numberOfBytesToAskFor =
+			65536; // (1 << 16); // This is the value in DarkLighting
+		uint32_t numberOfBytesReceived = 0;
+
+		char buffer[numberOfBytesToAskFor];
+
+		int ret = usbmuxd_recv_timeout(conn, (char *)&buffer,
+					       numberOfBytesToAskFor,
+					       &numberOfBytesReceived, 10);
+
+		if (ret == 0) {
+
+			if (getState() == State::Connecting) {
+				setState(State::Connected);
+			}
+
+			if (numberOfBytesReceived > 0) {
+				if (running) {
+					if (auto spt = delegate.lock()) {
+						// copy the char* bytes into the vector
+						auto vector =
+							std::vector<char>();
+						vector.insert(
+							vector.end(), buffer,
+							buffer +
+								numberOfBytesReceived);
```

Comment:
It seems like instead of doing this, you could write directly to the vector...

```suggestion
		const uint32_t numberOfBytesToAskFor =
			65536; // (1 << 16); // This is the value in DarkLighting
		uint32_t numberOfBytesReceived = 0;
		auto vector = std::vector<char>(numberOfBytesToAskFor);

		int ret = usbmuxd_recv_timeout(conn, vector.data(),
					       numberOfBytesToAskFor,
					       &numberOfBytesReceived, 10);

		if (ret == 0) {
			if (getState() == State::Connecting) {
				setState(State::Connected);
			}

			if (numberOfBytesReceived > 0) {
				vector.resize(numberOfBytesReceived);

				if (running) {
					if (auto spt = delegate.lock()) {
```

Reply:
Brilliant! 👍